src/engine/Drawing/CanvasDrawingSystem.ts

```diff
@@ -0,0 +1,107 @@
+  public readonly types = ['transform', 'canvas'] as const;
+  public systemType = SystemType.Draw;
+  public priority = -1;
+
+  private _ctx: CanvasRenderingContext2D;
+  private _camera: Camera;
+  private _engine: Engine;
+
+  public initialize(scene: Scene): void {
+    this._ctx = scene.engine.ctx;
+    this._engine = scene.engine;
+    this._camera = scene.camera;
+  }
+
+  public sort(a: Entity<TransformComponent | CanvasDrawComponent>, b: Entity<TransformComponent | CanvasDrawComponent>) {
+    return a.components.transform.z - b.components.transform.z;
+  }
+
+  public update(entities: Entity<TransformComponent | CanvasDrawComponent>[], delta: number) {
+    this._clearScreen();
+
+    let transform: TransformComponent;
+    let canvasdraw: CanvasDrawComponent;
+    const length = entities.length;
+    for (let i = 0; i < length; i++) {
+      if ((entities[i] as Actor).visible && !(entities[i] as Actor).isOffScreen) {
+        transform = entities[i].components.transform;
+        canvasdraw = entities[i].components.canvas;
+
+        this._ctx.save();
+        this._pushCameraTransform(transform);
+
+        this._ctx.save();
+        this._applyTransform(entities[i] as Actor);
+        canvasdraw.draw(this._ctx, delta);
+        this._ctx.restore();
+
+        this._popCameraTransform(transform);
+        this._ctx.restore();
+      }
+
+      if (this._engine.isDebug) {
+        this._ctx.save();
+        this._pushCameraTransform(transform);
+        this._ctx.strokeStyle = 'yellow';
+        (entities[i] as Actor).debugDraw(this._ctx);
+        this._popCameraTransform(transform);
+        this._ctx.restore();
+      }
+    }
+    if (this._engine.isDebug) {
+      this._ctx.save();
+      this._camera.draw(this._ctx);
+      this._camera.debugDraw(this._ctx);
+      this._ctx.restore();
+    }
+  }
+
+  private _applyTransform(actor: Actor) {
+    let parent = actor.parent;
+    while (parent) {
+      this._ctx.translate(parent.pos.x, parent.pos.y);
+      this._ctx.rotate(parent.rotation);
+      this._ctx.scale(parent.scale.x, parent.scale.y);
+      parent = parent.parent;
+    }
+
+    this._ctx.translate(actor.pos.x, actor.pos.y);
+    this._ctx.rotate(actor.rotation);
+    this._ctx.scale(actor.scale.x, actor.scale.y);
+  }
+
+  private _clearScreen(): void {
+    this._ctx.clearRect(0, 0, this._ctx.canvas.width, this._ctx.canvas.height);
+    this._ctx.fillStyle = this._engine.backgroundColor.toString();
+    this._ctx.fillRect(0, 0, this._ctx.canvas.width, this._ctx.canvas.height);
+  }
+
+  private _pushCameraTransform(transform: TransformComponent) {
+    // Establish camera offset per entity
+    if (transform.coordPlane === CoordPlane.World) {
+      this._ctx.save();
+      if (this._camera) {
+        this._camera.draw(this._ctx);
+      }
+    }
+  }
+
+  private _popCameraTransform(transform: TransformComponent) {
+    if (transform.coordPlane === CoordPlane.World) {
+      // Apply camera world offset
```

Comment:
is this applying the world offset to the camera, or applying a camera offset to the world?

Reply:
Good question, `_pushCameraTransform` applies the camera's transform to the graphics context which is world space. This comment here seems out of place in `_popCameraTransfrom` which restores the context back to screen space.

Reply:
I've updated the comments to give more clarity